url2df0f7c811_przemekgalazka_legacycode_tgz-pJ8-pl_cc_core_cmd_CommandJ8/src/ccproxy/AgentConnection.java

```diff
@@ -0,0 +1,1042 @@
+				out.println("+INFO Name '" + agentName + "', Extension '" + agentExten + "', AgentNumber '" + agentNumber + "', Type 'supervisor'");
+			} else {
+				out.println("+INFO Name '" + agentName + "', Extension '" + agentExten + "', AgentNumber '" + agentNumber + "', Type 'agent'");
+			}
+
+
+            if (offline){
+                out.close();
+                in.close();
+                out=new NullChannelWriter();
+            }
+
+			//wylistowanie rodzaju pauz dostępnych w systemie
+			printPauseTypesToAgent(out);
+
+			synchronized (p.agenci) {
+				p.agenci.add(this);
+			}
+
+            paused = p.c.pauseWhenStartup;
+
+			p.gadajAsterisk.sendMessage(this, "login", false);
+			
+			p.gadajAsterisk.sendMessage(this, "getallqueues", false);
+			
+			p.gadajAsterisk.sendMessage(this, "getqueues");
+				
+			//wylistowanie uprawnień
+			printPrivilegesToAgent(new Integer(agentNumber), out);
+			
+			
+			// FIXME dorobic catch do readline
+
+			last_ping = GregorianCalendar.getInstance().getTime();
+			last_db_activity = GregorianCalendar.getInstance().getTime();
+			mainThread = currentThread();
+            if (offline){
+             return;
+            }
+
+			agentConnectionPing = new AgentConnectionPing(this, mainThread);
+			agentConnectionPing.start();
+
+			
+			while (true) {
+				
+				if (licenseExpires!=null && GregorianCalendar.getInstance().getTime().after(licenseExpires)) {
+					p.c.licensedAgents="0";
+					p.log.info("License Expired");
+					out.close();
+					in.close();
+					return;
+				}
+							
+				selector.select();
+				Iterator selectedKeysIterator = selector.selectedKeys().iterator();
+
+				while (selectedKeysIterator.hasNext()) {
+					SelectionKey key = (SelectionKey) selectedKeysIterator.next();
+					selectedKeysIterator.remove();
+					if (!key.isValid())
+						continue;
+
+					if (key.isReadable()){
+						p.log.trace("Readable...");
+						
+						while ((outputLine=in.readLine())!=null){
+						if (outputLine!=null){
+							if (outputLine.equals("exit") || outputLine.equals("logout")) {
+								clearexit = true;
+								break;
+							}
+							if (outputLine.equals("ping")) {
+								logping.info("Ping fr ID " + idConnection + ", Agent: "
+										+ agentNumber + ", exten: " + agentExten + ", IP: "
+										+ ip + ":" + port);
+								out.println("+OK pong");
+								logping.info("Pong to ID " + idConnection + ", Agent: "
+										+ agentNumber + ", exten: " + agentExten + ", IP: "
+										+ ip + ":" + port);
+								last_ping = GregorianCalendar.getInstance().getTime();
+								continue;
+							}
+							p.gadajAsterisk.sendMessage(this, outputLine);
+							
+						}
+						}}
+					if (key.isWritable()){
+						p.log.trace("Writable...");
+						sChannel.register(selector, SelectionKey.OP_READ);
+						out.println(null);
+					}
+						
+					}
+				
+				
+				if (isInterrupted())
+					break;
+				try {
+					Thread.sleep(0);
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)